### ISSUE TITLE
fileUpload: handle invalid filename like url-encoded characters (%20) as validation error

### ISSUE THREAD

**Issue by spadovan** [primefaces/primefaces#12862]
### Describe the bug

If i try to upload a file whose filename contains url encoded characters like %20 (example 'attrezzatura+r%C3%A4te_signed.pdf') the component throws an exception (ValidatorException), but should display only an error message.

You can try this behaviour using showcase and the attached file
![attrezzatura+informatica+-+EDV-Ger%C3%A4te_signed](https://github.com/user-attachments/assets/fe6e5bb0-6884-4b49-b2a8-caf32c7ba3dc)

### PrimeFaces edition

Community

### PrimeFaces version

14.0.6

### JSF implementation

Mojarra

### JSF version

4

### Java version

17

### Browser(s)

all

**Comment by tandraschko** [comment]
your attachment doesnt work

**Comment by spadovan** [comment]
...sorry what does it mean ?

**Comment by melloware** [comment]
Can you put the attachment in a ZIP file so it retains the file name is what @tandraschko is saying.

**Comment by tandraschko** [comment]
ah, i thought you attached a project

where can we see on the showcase? plaese post exact steps

**Comment by spadovan** [comment]
Ok. go to showcase on the page of fileUpload, choose fileUpload basic or any other fileupload options. select the file with the wrong name, click on submit. Page not found page will appear, cauded by the excption thrown by the component.
I've uploaded the zip with the file.

Thanks
[file.zip](https://github.com/user-attachments/files/17702664/file.zip)

**Comment by melloware** [comment]
Yep I get this

```
jakarta.faces.validator.ValidatorException: File Error
        at org.primefaces.util.FileUploadUtils.validationError(FileUploadUtils.java:174)
        at org.primefaces.util.FileUploadUtils.requireValidFilename(FileUploadUtils.java:111)
        at org.primefaces.model.file.AbstractUploadedFile.<init>(AbstractUploadedFile.java:52)
        at org.primefaces.model.file.NativeUploadedFile.<init>(NativeUploadedFile.java:42)
        at org.primefaces.component.fileupload.NativeFileUploadDecoder.createUploadedFile(NativeFileUploadDecoder.java:70)
        at org.primefaces.component.fileupload.AbstractFileUploadDecoder.decodeSimple(AbstractFileUploadDecoder.java:84)
        at org.primefaces.component.fileupload.AbstractFileUploadDecoder.decode(AbstractFileUploadDecoder.java:60)
        at org.primefaces.component.fileupload.FileUploadRenderer.decode(FileUploadRenderer.java:59)
        at jakarta.faces.component.UIComponentBase.decode(UIComponentBase.java:500)
        at jakarta.faces.component.UIInput.decode(UIInput.java:646)
        at jakarta.faces.component.UIComponentBase.processDecodes(UIComponentBase.java:872)
```

**Comment by melloware** [comment]
our code specifically does this...

```java
// URL encoded characters like %20 are invalid
        Matcher encodedMatcher = ENCODED_CHARS_PAT.matcher(filename);
        if (encodedMatcher.find() ) {
            throw validationError( "Invalid filename: (" + filename + ") contains invalid character: " + encodedMatcher.group());
        }
```

**Comment by spadovan** [comment]
yes the problem is that this code is loaded in the costrunctor of the component, causing as exception to be thrown

**Comment by melloware** [comment]
Agreed.   So this issue isn't the `%` being blocked its to 500 error Page.

**Comment by spadovan** [comment]
yes, sorry for the misunderstanding...the problem is the error page thrown by the exception. An error message should be generated instead. thanks

**Comment by tandraschko** [comment]
yeah
we have no validator for it

**Comment by tandraschko** [comment]
or just move the requireValidFilename call into FileUpload#tryValidateFile instead of the in the decoder

**Comment by melloware** [comment]
Agreed I think us doing it in the constructor definitely feels like the wrong place now that I look at it.

**Comment by tandraschko** [comment]
@melloware 

now unit tests are failing on german machine:

[ERROR]   FileUploadUtilsTest.requireValidFilename_EncodedChars:379 expected: <Invalid filename: filename_with_%20 contains invalid character: %20> but was: <Ungültiger Dateiname: filename_with_%20 enthält ungültige Zeichen: %20>
[ERROR]   FileUploadUtilsTest.requireValidFilename_InvalidCharacters:419 expected: <Invalid filename: file:* contains invalid character: :> but was: <Ungültiger Dateiname: file:* enthält ungültige Zeichen: :>
[ERROR]   FileUploadUtilsTest.requireValidFilename_NotBlank:405 expected: <Filename cannot be empty or null> but was: <Dateiname darf nicht leer oder null sein>
[ERROR]   FileUploadUtilsTest.requireValidFilename_NotNull:393 expected: <Filename cannot be empty or null> but was: <Dateiname darf nicht leer oder null sein>

**Comment by SimonePadovan** [comment]
hi, is it possible to insert this fix in version 14.0.11 ? thanks

**Comment by tandraschko** [comment]
Its much more then a simple Bugfix, so we decided to keep it only in 15.x
